### PR TITLE
fix: close session in case of error to avoid blocked session

### PR DESCRIPTION
**Pull request by markusmueller**
…session on locking session storage (example unauth_action=deny)

### Description

This PR is fixing a bug where sessions are blocked until ttl is expired in case of an authentication error. For example unauthenticated session and `unauth_action='deny'` and locking session storage is configured.

Basically the same problem as in https://github.com/apache/apisix/pull/10788 but in case of error.

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by shreemaan-abhishek** [comment]
the solution looks good in first observation, let's wait for the CI to run then I will review this again.

**Comment by markusmueller** [comment]
@shreemaan-abhishek Is it possible that ci is flaky? The errors don't seem related to this PR to me.

File: t/plugin/openid-connect5.t [hunk 1]
```diff
@@ -138,3 +138,97 @@ __DATA__
...
+                                "bearer_only" : false,
+                                "timeout": 10,
+                                "introspection_endpoint_auth_method": "client_secret_post",
+                                "introspection_endpoint": "http://127.0.0.1:8080/realms/University/protocol/openid-connect/token/introspect",
+                                "set_access_token_header": true,
+                                "access_token_in_authorization_header": false,
+                                "set_id_token_header": true,
+                                "set_userinfo_header": true,
+                                "set_refresh_token_header": true,
+                                "unauth_action": "deny"
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect5.t:187]: are all these configurations required? We should consider using a minimal configuration.

Comment (markusmueller) [reply t/plugin/openid-connect5.t:178]: Yes, you are right, I have removed the properties which are not needed. Thank you!

Comment (shreemaan-abhishek) [reply t/plugin/openid-connect5.t:178]: what about these? 
image

Comment (markusmueller) [reply t/plugin/openid-connect5.t:167]: Also got minimal redis config now

**Comment by shreemaan-abhishek** [comment]
@starsz does this feature look good to you?

File: t/plugin/openid-connect5.t [hunk 2]
```diff
@@ -138,3 +138,77 @@ __DATA__
...
+                 [[{
+                        "plugins": {
+                            "openid-connect": {
+                                "discovery": "http://127.0.0.1:8080/realms/University/.well-known/openid-configuration",
+                                "realm": "University",
+                                "client_id": "course_management",
+                                "client_secret": "d1ec69e9-55d2-4109-a3ea-befa071579d5",
+                                "redirect_uri": "http://127.0.0.1:]] .. ngx.var.server_port .. [[/authenticated",
+                                "ssl_verify": false,
+                                "unauth_action": "deny"
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect5.t:167]: does unauth_action mean that all unauth reqs will be denied by default?

File: t/plugin/openid-connect5.t [hunk 3]
```diff
@@ -138,3 +138,77 @@ __DATA__
     }
 --- response_body_like
 hello world
+
+
+
+=== TEST 2: Call to route with locking session storage, no authentication and unauth_action 'deny' should not block subsequent requests on same session
+--- config
+    set $session_storage redis;
+    set $session_redis_uselocking               on;
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect5.t:147]: why are these two required?

File: t/plugin/openid-connect5.t [hunk 4]
```diff
@@ -138,3 +138,77 @@ __DATA__
...
+                )
+
+            local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/hello"
+
+            -- Make the final call to protected route WITHOUT cookie
+            local httpc = http.new()
+            local res, err = httpc:request_uri(uri, {method = "GET"})
+
+            -- Extract cookie which is not authenticated
+            local cookie_str = concatenate_cookies(res.headers['Set-Cookie'])
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect5.t:187]: I don't think the response would contain `set-cookie` header 🤔

Comment (shreemaan-abhishek) [reply t/plugin/openid-connect5.t:187]: okay my bad, I think the request will first be redirected to /authenticate

File: t/plugin/openid-connect5.t (hunk 2 above)
Comment (markusmueller) [reply t/plugin/openid-connect5.t:167]: yes

File: t/plugin/openid-connect5.t (hunk 3 above)
Comment (markusmueller) [reply t/plugin/openid-connect5.t:147]: To enable the locking session storage, thats what this bug is about

Comment (shreemaan-abhishek) [reply t/plugin/openid-connect5.t:147]: oops, well it was too obvious 😂

File: t/plugin/openid-connect5.t [hunk 5]
```diff
@@ -138,3 +138,77 @@ __DATA__
...
+            local function firstRequest()
+               local httpc = http.new()
+               httpc:request_uri(uri, {
+                        method = "GET",
+                        headers = {
+                            ["Cookie"] = cookie_str
+                        }
+                    })
+            end
+            ngx.thread.spawn(firstRequest)
```
Comment (starsz) [inline t/plugin/openid-connect5.t:199]: How can you ensure that the first request is arrived before the second one?
I think maybe you can give a return value of the `firstRequest` function , and use `ngx.thread.wait` to get the return value.

Comment (markusmueller) [reply t/plugin/openid-connect5.t]: @starsz Thank you for the suggestion, I have implemented a wait for the completion of the first request.

**Review by starsz — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]

**Pull request merged**